hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRejectHandler.java

```diff
@@ -0,0 +1,381 @@
+        final var nftTransfers = rejections.stream()
+                .filter(TokenReference::hasNft)
+                .map(rejection -> processNFTTokenRejection(rejection.nftOrThrow(), accountID, nftStore, tokenStore))
+                .collect(Collectors.groupingBy(Pair::left, Collectors.mapping(Pair::right, Collectors.toList())))
+                .entrySet()
+                .stream()
+                .map(tokenIdNFTPair -> createNftTransfer(tokenIdNFTPair.getKey(), tokenIdNFTPair.getValue()))
+                .toList();
+
+        final var tokenTransferLists = new ArrayList<TokenTransferList>();
+        tokenTransferLists.addAll(fungibleTransfers);
+        tokenTransferLists.addAll(nftTransfers);
+        return tokenTransferLists;
+    }
+
+    /**
+     * Processes a single NFT rejection by performing validations and converting it to a Pair of TokenID & NftTransfer.
+     *
+     * @param nftID The NftID of the NFT being rejected.
+     * @param rejectingAccountID The AccountID of the owner rejecting the NFT.
+     * @param nftStore Access to writable NFT data.
+     * @param tokenStore Access to writable token data.
+     */
+    private Pair<TokenID, NftTransfer> processNFTTokenRejection(
+            @NonNull final NftID nftID,
+            @NonNull final AccountID rejectingAccountID,
+            @NonNull final WritableNftStore nftStore,
+            @NonNull final WritableTokenStore tokenStore) {
+        final var nft = getIfUsable(nftID, nftStore);
+        final var token = getIfUsable(nftID.tokenIdOrThrow(), tokenStore);
+        final var tokenTreasury = token.treasuryAccountIdOrThrow();
+        validateTrue(!tokenTreasury.equals(rejectingAccountID), ACCOUNT_IS_TREASURY);
+        validateTrue(nft.hasOwnerId() && rejectingAccountID.equals(nft.ownerId()), INVALID_OWNER_ID);
+
+        return Pair.of(token.tokenId(), nftTransfer(rejectingAccountID, tokenTreasury, nftID.serialNumber()));
+    }
+
+    /**
+     * Processes a single fungible token rejection by performing validations and builds the transfer for the rejected
+     * token.
+     *
+     * @param rejection The token reference detailing the rejection.
+     * @param rejectingAccount The Account rejecting the token.
+     * @param tokenStore Access to writable token data.
+     * @param relStore Access to writable token relations.
+     */
+    private TokenTransferList processFungibleTokenRejection(
+            @NonNull final TokenReference rejection,
+            @NonNull final Account rejectingAccount,
+            @NonNull final WritableTokenStore tokenStore,
+            @NonNull final WritableTokenRelationStore relStore) {
+        final var accountID = rejectingAccount.accountIdOrThrow();
+        final var token = getIfUsable(rejection.fungibleTokenOrThrow(), tokenStore);
+        final var tokenId = token.tokenIdOrThrow();
+        final var tokenRelation = getIfUsable(accountID, tokenId, relStore);
```

Comment:
Wouldn't this fail if the token is paused or frozen ?

Reply:
It would indeed fail.